File: src/main/kotlin/cn/bakamc/folia/item/SpecialItemManager.kt
package cn.bakamc.folia.item

import cn.bakamc.folia.db.table.SpecialItem
import cn.bakamc.folia.service.SpecialItemService
import cn.bakamc.folia.util.launch
import cn.bakamc.folia.util.logger
import moe.forpleuvoir.nebula.common.api.Initializable
import java.util.concurrent.ConcurrentHashMap

object SpecialItemManager : Initializable {

    private var itemCache: MutableMap<String, SpecialItem> = ConcurrentHashMap()

    override fun init() {
        launch {
            SpecialItemService.getSpecialItems().forEach {
                itemCache[it.key] = it
            }
            logger.info("特殊物品加载完成")
        }
    }

    fun onDisable() {
        itemCache.clear()
    }

    fun getCachedItem(key: String): SpecialItem? {
        return itemCache[key].apply {
            if (this == null) {
                launch {
                    SpecialItemService.getItemByKey(key)?.let {
                        itemCache[it.key] = it
                    }
                }
            }
        }
    }

    fun getCache(): Map<String, SpecialItem> {
        return itemCache
    }

    suspend fun put(specialItem: SpecialItem): Boolean {
        SpecialItemService.inertOrUpdate(specialItem)?.let {
            itemCache[it.key] = it
            return true
        }
        return false
    }

    suspend fun remove(key: String): SpecialItem? {
        return SpecialItemService.delete(key).takeIf { it > 0 }?.let {
            itemCache.remove(key)
        }
    }

    fun specifyType(keys: Set<String>): Map<String, SpecialItem> {
        return buildMap {
            keys.forEach {
                itemCache[it]?.let { specialItem ->
                    put(specialItem.key, specialItem)
                }
            }
        }
    }

}

File: bakamc-folia/src/main/scala/cn/bakamc/folia/functional/specialitem/SpecialItemManager.scala
package cn.bakamc.folia.functional.specialitem

import cn.bakamc.folia.BakaMC
import cn.bakamc.folia.database.services.SpecialItemService
import cn.bakamc.folia.database.table.SpecialItem
import moe.forpleuvoir.nebula.common.api.Initializable

import java.util.concurrent.ConcurrentHashMap
import scala.concurrent.Await
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration.{DurationInt, FiniteDuration}
import scala.jdk.CollectionConverters.*
import language.implicitConversions

object SpecialItemManager extends Initializable {

  private[specialitem] val timeoutDuration: FiniteDuration = 20.second

  private[specialitem] val cache = ConcurrentHashMap[String, SpecialItem]()

  private[specialitem] def logger = BakaMC.logger

  private[specialitem] def service = SpecialItemService

  override def initialization(): Unit = {
    cache.clear()
    try {
      Await.result(service.getSpecialItems, timeoutDuration)
        .foreach { item =>
          cache.put(item.id, item)
        }
      logger.info("特殊物品加载完成 {}", cache.size)
    } catch {
      case e: Throwable => logger.error("特殊物品加载失败", e)
    }
  }

  def getCache: Map[String, SpecialItem] = cache.asScala.toMap

  def getCachedItem(id: String): Option[SpecialItem] = Option(cache.get(id))

  def update(item: SpecialItem)(result: Boolean => Unit): Unit = {
    service.inertOrUpdate(item).foreach {
      case r if r > 0 =>
        cache.put(item.id, item)
        result(true)
      case _ => result(false)
    }
  }

  def delete(id: String)(result: Boolean => Unit): Unit = {
    service.delete(id).foreach {
      case r if r > 0 =>
        cache.remove(id)
        result(true)
      case _ => result(false)
    }
  }

  def specifyType(ids: Set[String]): Map[String, SpecialItem] = {
    val builder = Map.newBuilder[String, SpecialItem]
    ids.foreach { id =>
      getCachedItem(id).foreach { item =>
        builder.addOne(id -> item)
      }
    }
    builder.result()
  }

}
